File: GasProject Server/Sources/World/Objects/Ghost.cpp
#include "Ghost.hpp"

#include "Control.hpp"

#include <World/Map.hpp>
#include <World/Tile.hpp>

Ghost::Ghost() : hostControl(nullptr)
{
	layer = 80;
	sprite = "ghost";
	name = "Ghost";
	density = false;
	invisibility = 1;
	seeInvisibleAbility = 1;
}

void Ghost::MoveZ(int order) {
	if (!order) {
		return;
	}
	Tile *tile = GetTile();
	Tile *dest_tile = tile->GetMap()->GetTile(tile->GetPos() + rpos(0,0,order));
	if (dest_tile) {
		dest_tile->MoveTo(this);
	}
}

void Ghost::SetHostControl(Control *control) {
	hostControl = control;
}

Control *Ghost::GetHostControl() const { return hostControl; }

File: GameLogic/Objects/Creatures/Ghost.py
from Objects.Creature import Creature
from Engine.World import CreateObject, Object
from Engine_Geometry import Vec2i

class Ghost(Creature):
	def __init__(self):
		Creature.__init__(self)
		self.layer = 80
		self.name = "Python Ghost"
		self.sprite = "ghost"
		self.density = False
		self.seeInvisibleAbility = 1
		self.invisibility = 1

	@property
	def hostControl(self):
		return self._hostControl

	@hostControl.setter
	def hostControl(self, value):
		self._hostControl = value

	# IHasOrgans methods
	def CreateOrgans(self):
		pass


def Ghostize(player):
	mob = player.control.GetOwner()

	if not isinstance(mob, Ghost):
		tile = player.control.GetOwner().tile
		ghost = CreateObject("Objects.Creatures.Ghost", tile)
		ghost.hostControl = player.control
		player.control = ghost.GetComponent("Control")
	else:
		ghost = mob
		player.control = ghost.hostControl
		ghost.Delete()
